### PR TITLE
Add jax.image.resize.

### PR DESCRIPTION
**Pull request by hawkinsp**
This is a port of tf.image.resize() and the ScaleAndTranslate operator.

While I don't expect this implementation to be particularly fast, it is a useful generic implementation to which we can add optimized special cases as the need arises.

### CODE REVIEW

File: jax/image/scale.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+      np.not_equal(input_shape, output_shape) |
+      np.not_equal(scale, 1) |
+      np.not_equal(translate, 0))[0]
+  if len(spatial_dims) == 0:
+    return x
+  output_spatial_shape = tuple(np.array(output_shape)[spatial_dims])
+  indices = []
+  dim_weights = []
+  slice_shape = list(input_shape)
+  in_names = string.ascii_lowercase[:len(output_shape) + len(spatial_dims)]
```
Comment (shoyer) [inline jax/image/scale.py:97]: `einsum` also supports an alternative syntax that doesn't require string parsing. That might be slightly cleaner here?
```
a = np.arange(60.).reshape(3,4,5)
b = np.arange(24.).reshape(4,3,2)
np.einsum('ijk,jil->kl', a, b)
array([[4400., 4730.],
       [4532., 4874.],
       [4664., 5018.],
       [4796., 5162.],
       [4928., 5306.]])
np.einsum(a, [0,1,2], b, [1,0,3], [2,3])  # equivalent to the above
array([[4400., 4730.],
       [4532., 4874.],
       [4664., 5018.],
       [4796., 5162.],
       [4928., 5306.]])
```

(not really important given that this is already working!)

File: jax/image/scale.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    indices.append(starts.astype(np.int32))
+    dim_weights.append(weights.astype(x.dtype))
+    einstr += ",{}{}".format(in_names[len(output_shape) + i], in_names[d])
+    out_names[d] = in_names[len(output_shape) + i]
+  index = lax.concatenate(indices, len(output_spatial_shape))
+  dnums = lax.GatherDimensionNumbers(
+      offset_dims=tuple(range(len(output_shape))),
+      collapsed_slice_dims=(),
+      start_index_map=tuple(spatial_dims))
+  out = lax.gather(x, index, dnums, slice_shape)
```
Comment (shoyer) [inline jax/image/scale.py:116]: might be clearer to use numpy style indexing with brackets [], at least for someone who doesn't know XLA quite as well :)

**Comment by shoyer** [comment]
I would suggest also adding an explicit test for gradients. Even though there are no custom gradients defined here, it's possible to hit edge cases where gradients fail due to rounding, like in https://github.com/google/jax/issues/3024

File: jax/image/scale.py (hunk 2 above)
Comment (hawkinsp) [reply jax/image/scale.py:116]: Actually this is a case where I think I cannot do that. For example, for a bicubic resize (without antialiasing) this is gathering 4x4 patches in a single gather, but I don't see a way to gather windows with NumPy's indexing, which only really seems to support gathering single elements or entire axes.

File: jax/image/scale.py (hunk 1 above)
Comment (hawkinsp) [reply jax/image/scale.py:97]: Ah, I was looking for that. It's completely undocumented, I think? https://numpy.org/doc/stable/reference/generated/numpy.einsum.html

That said I felt dirty manipulating strings, so I will try this.

Comment (shoyer) [reply jax/image/scale.py:97]: It's definitely under-documented, but it's nicer for programmatic use

(and it works in JAX, too, because we copied the argument parsing logic from NumPy)

**Comment by hawkinsp** [comment]
> I would suggest also adding an explicit test for gradients. Even though there are no custom gradients defined here, it's possible to hit edge cases where gradients fail due to rounding, like in #3024

Done. I found 2nd-order gradients required a larger `eps` to pass, but these functions are linear so that seems acceptable to me.

**Comment by mattjj** [comment]
Yeah I think big `eps` is better for linear functions (not just acceptable) because it lets us drive numerical error down without introducing any approximation error.

File: jax/image/scale.py (hunk 1 above)
Comment (mattjj) [reply jax/image/scale.py]: For posterity, we implemented both forms of argument parsing "by hand" in [Autograd's einsum vjp](https://github.com/HIPS/autograd/blob/c6f630a5ec18bd30f1485bc0dbbccb8664c77510/autograd/numpy/numpy_vjps.py#L591-L626), and the non-string form was super easy to handle. I suspect it's meant for easy programmatic manipulation, whereas the strings can be easier for humans. (I do love how undocumented it is though, except for one sentence in the docstring!)

File: jax/image/scale.py (hunk 2 above)
Comment (shoyer) [reply jax/image/scale.py:114]: You can do this by adding an extra dimension to the array indexers, e.g.,
```
In [13]: image = np.zeros((50, 60))

In [14]: x, y = np.meshgrid(np.arange(40), np.arange(50), indexing='ij')

In [15]: image[x, y].shape
Out[15]: (40, 50)

In [16]: image[np.stack([x, x + 1], axis=-1), np.stack([y, y + 1], axis=-1)].shape
Out[16]: (40, 50, 2)
```

Comment (hawkinsp) [reply jax/image/scale.py:114]: This isn't quite the same thing once lowered to `lax.gather`. Your formulation does gathers of multiple indices, whereas the formulation I'm using here does a single gather of a continuous span. I don't know whether it makes a difference in practice, but it certainly seems like something the compiler could benefit from knowing. And it makes the indices array smaller!

File: jax/image/scale.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+class ResizeMethod(enum.Enum):
+  LINEAR = 1
+  LANCZOS3 = 2
+  LANCZOS5 = 3
+  CUBIC = 4
+
+  @staticmethod
+  def from_string(s: str):
+    if s in ['linear', 'bilinear', 'trilinear']:
```
Comment (johnpjf) [inline jax/image/scale.py:128]: could add triangle here as an option.

File: jax/image/scale.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                   antialias: bool) -> Tuple[np.ndarray, np.ndarray]:
+  """
+  Computes the locations and weights of the spans of a 1D input image.
+  Returns:
+    A `(starts, weights)` tuple of ndarrays, where `starts` has shape
+    `[output_size]` and `weights` has shape `[output_size, span_size]`.
+  """
+  radius, kernel_fn = kernel
+  inv_scale = 1. / scale
+  kernel_scale = max(inv_scale, 1.) if antialias else 1.
```
Comment (johnpjf) [inline jax/image/scale.py:65]: might be worth a comment here:
When downsampling the kernel should be scaled since we want to low pass
filter and interpolate, but when upsampling it should not be since we only
want to interpolate.

File: jax/image/scale.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+import math
+from typing import Callable, Sequence, Tuple, Union
+
+from jax import lax
+from jax import numpy as jnp
+import numpy as np
+
+
+def _lanczos_kernel(radius: float):
+  def kernel(x: np.ndarray) -> np.ndarray:
```
Comment (johnpjf) [inline jax/image/scale.py:25]: Just interested, why is this numpy and not jnp?

File: jax/image/scale.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+  kernel_pos = (span_start - sample_f)[:, None] + np.arange(span_size) + 0.5
+  weight = kernel_fn(kernel_pos / kernel_scale)
+  total_weight_sum = np.sum(weight, axis=1, keepdims=True)
+  weights = np.where(
+      np.abs(total_weight_sum) > 1000. * np.finfo(np.float32).min,
+      weight / total_weight_sum, 0)
+  return span_start.astype(np.int32), weights
+
+
+def _scale_and_translate(x, output_shape, scale, translate, kernel,
```
Comment (johnpjf) [inline jax/image/scale.py:80]: I think we should expose this function too, it's useful sometimes - for example you can use it to pull a region out of an image, or to mimic the "align corners" mode of TF's old resize which is needed in certain Unets I think.

File: tests/image_test.py [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+  @parameterized.named_parameters(jtu.cases_from_list(
+       {"testcase_name": "_shape={}_target={}_method={}_antialias={}".format(
+          jtu.format_shape_dtype_string(image_shape, dtype),
+          jtu.format_shape_dtype_string(target_shape, dtype), method,
+          antialias),
+        "dtype": dtype, "image_shape": image_shape,
+        "target_shape": target_shape,
+        "method": method, "antialias": antialias}
+       for dtype in [np.float32]
```
Comment (johnpjf) [inline tests/image_test.py:58]: do we want to test half float?

File: tests/image_test.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+          jtu.format_shape_dtype_string(target_shape, dtype), method),
+        "dtype": dtype, "image_shape": image_shape, "target_shape": target_shape,
+        "method": method}
+       for dtype in inexact_dtypes
+       for image_shape, target_shape in [
+         ([3, 1, 2], [6, 1, 4]),
+         ([1, 3, 2, 1], [1, 6, 4, 1]),
+       ]
+       for method in ["linear", "lanczos3", "lanczos5", "cubic"]))
+  def testResizeUp(self, dtype, image_shape, target_shape, method):
```
Comment (johnpjf) [inline tests/image_test.py:125]: Is there a testResizeDown, or is that implicit in the other tests.

**Review by johnpjf — APPROVED** [review]
This is awesome, can we scrap the TF C++ code and replace with similar code and we'll get a GPU version?

File: jax/image/scale.py (hunk 5 above)
Comment (hawkinsp) [reply jax/image/scale.py:25]: The span calculation is a function of the scale and the input shapes, and both are fixed at compile time in the current implementation. So we can use classic NumPy to calculate the weights; no need to lift it into JAX and, say, run it on the GPU.

File: jax/image/scale.py (hunk 6 above)
Comment (hawkinsp) [reply jax/image/scale.py:80]: Happy to do so, but let's do that in another PR. (I'm mostly concerned about needing to write tests, etc. and I'd like to get this in to unblock users.)

File: tests/image_test.py (hunk 8 above)
Comment (hawkinsp) [reply tests/image_test.py:125]: No, not at the moment. As you may have noticed, I ported `testResizeUp` from TensorFlow (ok in this case since it is Google-authored).

However I found that the corresponding `testResizeDown` actually seems not to apply to any of these resizes because the TF test isn't actually testing any methods except for nearest-neighbor. The logic here:
https://github.com/tensorflow/tensorflow/blob/9065899e9252d5d9472b45d5a3dbecfb8b039117/tensorflow/python/ops/image_ops_test.py#L2511
only enables that test for resizes that run on GPU, but only nearest neighbor is implemented on GPU.

Since I didn't have a great deal of confidence in producing correct golden outputs, I thought it safer just to test against PIL and TensorFlow. Note that those tests *do* test downsampling because they make a random selection of input and output shapes.

File: tests/image_test.py (hunk 7 above)
Comment (hawkinsp) [reply tests/image_test.py]: Done.

**Comment by hawkinsp** [comment]
> This is awesome, can we scrap the TF C++ code and replace with similar code and we'll get a GPU version?

Yes and no. This implementation will run on all platforms XLA supports, but it may or may not outperform a handwritten version. But I think that's a good starting point! At least this way you have, say, a 7D Lanczos-5 resize implementation, and we can optimize it if needed. Researchers often appreciate generality!

I think the key question for optimization here is whether XLA is able to fuse the gather with the matrix multiplications. On CPU and GPU it probably will not at the moment, whereas on TPU it may. Improving that is certainly feasible with more work on XLA.

In cases where the input and kernel sizes have well-behaved GCD relationships to each other, we could use XLA's convolution operator, which should perform well on all platforms. We implemented bilinear resizes that way in the TF/XLA bridge: https://github.com/tensorflow/tensorflow/blob/9065899e9252d5d9472b45d5a3dbecfb8b039117/tensorflow/compiler/tf2xla/kernels/image_resize_ops.cc#L36 although that implementation can lead to excessively large convolution kernels if people do things like 256 -> 299 upscaling (thanks, GAN folks). I may add that as an optimization in the future, but I think we need the "gather and matmul" implementation as well to handle those cases.
